design/service-apis-implementation.md

```diff
@@ -0,0 +1,253 @@
+Currently, Contour provides "TCP Proxying" that allows the forwarding of TLS streams based on the SNI information, which is precisely what the Service APIs TLSRoute object is for.
+If Project Contour (the organisation) does add support for TCP and UDP forwarding, it will not be in the `projectcontour/contour` repo, but will be a separate repo.
+
+## Goals
+- Define a data model and implementation for Contour's service-apis support.
+- Layer 7 support only, which means HTTPRoutes, TLSRoutes only.
+
+## Non Goals
+- No TCPRoute or UDPRoute support, that is, no support for arbitrary TCP/UDP forwarding.
+
+
+## High-Level Design
+
+### Service APIs implementation surface area
+
+Contour's support for the Service APIs will include support for the HTTPRoute and TLSRoute objects only, not any other type of Route - this includes no TCPRoute or UDPRoute support.
+Contour is a layer 7 ingress controller, and the layer 4 load balancing implies by TCPRoute and UDPRoute is out of scope for this tool (that is, `projectcontour/contour`).
+Project Contour (the organisation) may investigate a Service-APIs based layer 4 solution in the future, but that effort will not be in `projectcontour/contour`.
+
+### Contour Gateway model
+
+Contour considers a Gateway to describe a single Envoy deployment, deployed and managed by something outside of itself.
+Contour expects to be supplied with the full name (name and namespace) of a single Gateway, which it will watch and update the status of.
+Inside Contour will merge Listeners within its Gateway. (See the detailed design for the exact rules Contour will use for this.)
+
+### Interrelated watches
+
+The Service APIs are a set of interrelated Kubernetes objects, where a change in one object can mean that the scope of objects Contour is interested in will change.
+Because of this, Contour will watch all events associated with its named Gateway, HTTPRoute, TLSRoute, and BackendPolicy objects, and filter the objects it takes action on internally, rather than using a filtered watch (as those are costly to set up and tear down).
+
+Contour does not watch a GatewayClass, as its deployment model expects a Contour to be part of the implementation for a *particular* Gateway.
+
+### Combining Service APIs with other configuration
+
+When it calculates the Envoy configuration using the DAG, Contour layers different types of configuration in order.
+So, currently, Ingress is overwritten by HTTPProxy, so if an Ingress and a HTTPProxy specify the exact same route, the HTTPProxy will win.
+
+Once we have the Service-APIs available as well, we have to choose the order.
+In this design, I suggest having the order be Ingress is overwritten by HTTPProxy, is overwritten by the service-apis.
+I could see reversing HTTPProxy and the Service APIs here, but I think this acknowledges that the Service APIs are really an evolution of the ideas in HTTPProxy, and will probably end up being the community standard.
```

Comment:
Is it a goal to eventually sunset `HTTPProxy` and have service APIs be the only API for Contour? If not yet a goal, do we think it might be some day? Just curious how others are thinking about this, and whether we expect to support multiple APIs indefinitely or if we hope to converge on one.

Reply:
I'd like to think that, if the Service APIs had full feature parity with HTTPProxy, and our users were willing to move, we would remove the CRD.

I don't think either of those is likely though, so what I'm trying to do is behave as though that might be the case someday. Practically, I don't think it will be, but I'm not convinced of the benefit of having our own CRD if the above two conditions are fulfilled.